xlators/cluster/afr/src/afr-open.c

```diff
@@ -271,11 +274,204 @@ afr_fd_ctx_need_open(fd_t *fd, xlator_t *this, unsigned char *need_open)
+                         local->loc.path, priv->children[i]->name);
+            STACK_WIND_COOKIE(frame, afr_openfd_fix_open_cbk, (void *)(long)i,
+                              priv->children[i],
+                              priv->children[i]->fops->opendir, &local->loc,
+                              local->fd, NULL);
+        } else {
+            gf_msg_debug(this->name, 0,
+                         "opening fd for file %s on subvolume %s",
+                         local->loc.path, priv->children[i]->name);
+
+            STACK_WIND_COOKIE(frame, afr_openfd_fix_open_cbk, (void *)(long)i,
+                              priv->children[i], priv->children[i]->fops->open,
+                              &local->loc, local->fd_ctx->flags & (~O_TRUNC),
+                              local->fd, NULL);
+        }
+        if (!--need_open_count)
+            break;
+    }
+    return 0;
+
+out:
+    AFR_STACK_DESTROY(frame);
+    return 0;
+}
+
+static int
+afr_strict_lock_getlk_cbk(call_frame_t *frame, void *cookie, xlator_t *this,
+                          int32_t op_ret, int32_t op_errno,
+                          struct gf_flock *lock, dict_t *xdata)
+{
+    afr_local_t *local = frame->local;
+    afr_private_t *priv = NULL;
+    unsigned char *success_replies = NULL;
+    gf_boolean_t locks_held = _gf_false;
+    int ret = -1;
+    int call_count = 0;
+    int i = (long)cookie;
+
+    priv = this->private;
+    local->replies[i].valid = 1;
+    local->replies[i].op_ret = op_ret;
+    local->replies[i].op_errno = op_errno;
+    if (op_ret != 0) {
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, AFR_MSG_LK_HEAL_DOM,
+               "Failed getlk for %s", uuid_utoa(local->fd->inode->gfid));
+    } else {
+        local->cont.lk.getlk_rsp1[i] = *lock;
+    }
+
+    LOCK(&frame->lock);
+    {
+        if (!local->strict_locks) {
+            ret = dict_get_int32(xdata, "strict-locks", &local->strict_locks);
+            if (ret)
+                gf_msg(this->name, GF_LOG_ERROR, -ret, AFR_MSG_DICT_GET_FAILED,
+                       "Failed to get value of strict-locks setting");
+        }
+    }
+    UNLOCK(&frame->lock);
+
+    call_count = afr_frame_return(frame);
+
+    if (call_count)
+        return 0;
+
+    if (!local->strict_locks) {
+        afr_do_fix_open(frame, this);
+        return 0;
+    }
+
+    for (i = 0; i < priv->child_count; i++) {
+        if (!local->replies[i].valid || local->replies[i].op_ret != 0) {
+            continue;
+        }
+        if (local->cont.lk.getlk_rsp1[i].l_type == F_UNLCK) {
+            continue;
+        } else if (local->cont.lk.getlk_rsp1[i].l_type == F_WRLCK) {
+            locks_held = _gf_true;
+            break;
+        }
+    }
+
+    success_replies = alloca0(priv->child_count);
+    afr_fill_success_replies(local, priv, success_replies);
+    if (AFR_COUNT(success_replies, priv->child_count) == 0) {
+        goto out;
+    }
```

Comment:
In cases where we get op_ret != 0/invalid reply from all the bricks, at that time we will not have the locks_held value set to true and that will lead to re-opening of fd. To avoid that, we need this check.

Reply:
Yes, locks held will be false and it will go ahead and re-open. To avoid that, I am checking for success replies here.

Reply:
This was added to handle a case like below:
1. fd1 is opened and lock was taken on it when brick1 was down
2. Bricks 2 & 3 disconnect & reconnects at some point
2. Now brick1 needs re-open and bricks 2 & 3 are returning EBADFD
3. Since lock is not held on the any of the bricks, a new lock request on fd2 will be allowed
4. Now if we try to write some data on fd1, it will go ahead and re-open the fd on brick1 if we do not check for success replies. This is because all the replies had non-zero op-ret and locks_held will be false at this point. So it will go ahead with re-open & write the data even though it does not have a lock.

In the current ode all these are removed and check for active locks is done in the below layer.